packages/backend/src/tokens/types.ts

```diff
@@ -10,5 +10,60 @@ export type AuthenticateRequestOptions = {
   signUpUrl?: string;
   afterSignInUrl?: string;
   afterSignUpUrl?: string;
+  organizationSyncOptions?: OrganizationSyncOptions;
   apiClient?: ApiClient;
 } & VerifyTokenOptions;
+
+/**
+ * Defines the options for syncing an organization or personal workspace state from the URL to the clerk session.
+ * Useful if the application requires the inclusion of a URL that indicates that a given clerk organization
+ * (or personal workspace) must be active on the clerk session.
+ *
+ * If a mismatch between the active organization on the session and the organization as indicated by the URL is
+ * detected, an attempt to activate the given organization will be made.
+ *
+ * WARNING: If the activation cannot be performed, either because an organization does not exist or the user lacks
+ * access, then the active organization on the session will not be changed (and a warning will be logged). It is
+ * ultimately the responsibility of the page to verify that the resources are appropriate to render given the URL,
+ * and handle mismatches appropriately (e.g. by returning a 404).
+ */
+export type OrganizationSyncOptions = {
+  /**
+   * URL patterns that are organization-specific and contain an organization ID or slug as a path token.
+   * If a request matches this path, the organization identifier will be extracted and activated before rendering.
+   *
+   * WARNING: If the organization cannot be activated either because it does not exist or the user lacks access,
+   * organization-related fields will be set to null. The server component must detect this and respond
+   * with an appropriate error (e.g., notFound()).
+   *
+   * If the route also matches the personalWorkspacePatterns, the personalWorkspacePattern takes precedence.
+   *
+   * Must have a path token named either ":id" (matches a clerk organization ID) or ":slug" (matches a clerk
+   * organization slug).
+   *
+   * Common examples:
+   * - ["/orgs/:slug", "/orgs/:slug/(.*)"]
+   * - ["/orgs/:id", "/orgs/:id/(.*)"]
+   * - ["/app/:any/orgs/:slug", "/app/:any/orgs/:slug/(.*)"]
+   */
+  organizationPatterns?: Array<Pattern>;
+
+  /**
+   * URL patterns for resources in the context of a clerk personal workspace (user-specific, outside any organization).
+   * If the route also matches the organizationPattern, this takes precedence.
+   *
+   * Common examples:
+   * - ["/user", "/user/(.*)"]
+   * - ["/user/:any", "/user/:any/(.*)"]
+   */
+  personalWorkspacePatterns?: Array<Pattern>;
+};
```

Comment:
Maybe there is a benefit of redefining those for the `nextjs` package and take advantage of our  [RouteMatcherWithNextTypedRoutes](https://github.com/clerk/javascript/blob/7f873916ed95ed852acfc82680fe71fbc79ec662/packages/nextjs/src/server/routeMatcher.ts#L9)

Reply:
And just to double check here, we don't wanna use the `createRouteMatcher` pattern here, because we are not expecting for folks to use more than one path ? Although the name of the properties hints that you can. If not, I think supporting the route matcher here makes sense.

cc @nikosdouvlis

Reply:
Good idea! I didn't realize we already have [RouteMatcherParams](https://github.com/search?q=repo%3Aclerk%2Fjavascript+%22export+type+RouteMatcherParam%22&type=code), which serves the same purpose.

It looks like there's some duplication between the astro and nextjs RouteMatcherParams. My inclination here is to make the astro version (which doesn't depend on anything astro-specific) shared, and import that for use here. But let me know if you think another path would be better!

r/e createRouteMatcher - We could take a createRouteMatcher-returned function here too, but my inclination is to start with just accepting the pattern syntax for simplicity and expand the the more complex types if necessary in the future - I explored that a bit here (internal only): https://www.notion.so/Sync-Org-from-URL-to-Session-via-Middleware-f41f00865390480ab2279391c5625b27?d=f3738cb7f62c48bb90f8924c653a83a9&pvs=4#c0b57bf937694dd6bf510aeb4dde12af. Again, differing opinions welcome!

Reply:
Oh you know, digging in a bit further, I think the main difference here between this and `RouteMatcherWithNextTypedRoutes`/`RouteMatcherRoutes` is that they're geared at returning true/false, not pulling information out of the path. For the organization pattern, we need it to include a group (which I have here as `:slug` or `:id`). I don't see a clear way to extend the RouteMatcherParams to encompass that without making the type too complex for my taste. I could have the personal workspace pattern use RouteMatcherParams, but I think the mixed types would be worse DX. 

And to answer your question:
>  we are not expecting for folks to use more than one path ?

I am expecting more than one path! It's very tricky to get most practical examples down to just one path expression, so I think allowing multiples is the way to go.